nixos/tests/incus/container.nix

```
@@ -0,0 +1,113 @@
+  container-image-rootfs = releases.lxdContainerImage.${pkgs.stdenv.hostPlatform.system};
+in
+{
+  name = "incus-container";
+
+  meta.maintainers = with lib.maintainers; [ adamcstephens ];
+
+  nodes.machine = { ... }: {
+    virtualisation = {
+      diskSize = 4096;
+
+      # Since we're testing `limits.cpu`, we've gotta have a known number of
+      # cores to lean on
+      cores = 2;
+
+      # Ditto, for `limits.memory`
+      memorySize = 512;
+
+      incus.enable = true;
+    };
+  };
+
+  testScript = ''
+    def instance_is_up(_) -> bool:
+      status, _ = machine.execute("incus exec container --disable-stdin --force-interactive /run/current-system/sw/bin/true")
+      return status == 0
+
+    machine.wait_for_unit("incus.service")
+
+    # no preseed should mean no service
+    machine.fail("systemctl status incus-preseed.service")
+
+    machine.succeed("incus admin init --minimal")
+
+    machine.succeed(
+        "incus image import ${container-image-metadata}/*/*.tar.xz ${container-image-rootfs}/*/*.tar.xz --alias nixos"
+    )
+
+    with subtest("Container can be managed"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+          retry(instance_is_up)
+        machine.succeed("echo true | incus exec container /run/current-system/sw/bin/bash -")
+        machine.succeed("incus delete -f container")
+
+    with subtest("Container is mounted with lxcfs inside"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        ## ---------- ##
+        ## limits.cpu ##
+
+        machine.succeed("incus config set container limits.cpu 1")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "1"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
+
+        machine.succeed("incus config set container limits.cpu 2")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "2"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
+
+        ## ------------- ##
+        ## limits.memory ##
+
+        machine.succeed("incus config set container limits.memory 64MB")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "MemTotal:          62500 kB"
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
```

Let's add a bit more fault tolerance, and not test the indentation

```suggestion
        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
        assert meminfo_bytes == "62500 kB", "Wrong amount of memory reported from /proc/meminfo"
```